### PR TITLE
Chore/143 migrate plug

### PR DESCRIPTION
**Pull request by aliXsed**
Migrating the latest develop branch of Plug into cennznet develop.

I would like the reviewers to look at the following commits:
https://github.com/cennznet/cennznet/commit/d4a5d4911631e35c8daf029c77a17b9b8eaf8c4d
https://github.com/cennznet/cennznet/commit/d0dcece675840e62f61d1994878773cc076648a5

Note: before we actually merge this branch we will change all the references to Plug develop to refer to Plug rc3 instead.

### CODE REVIEW

File: crml/transaction-payment/Cargo.toml [hunk 1]
```diff
@@ -3,20 +3,25 @@ name = "crml-transaction-payment"
 version = "1.0.0"
 authors = ["Centrality Developers <support@centrality.ai>"]
 edition = "2018"
+license = "GPL-3.0"
+homepage = "https://cennznetdocs.com"
```
Comment (jordy25519) [inline crml/transaction-payment/Cargo.toml]: Can remove this

File: executor/Cargo.toml [hunk 2]
```diff
@@ -2,34 +2,56 @@
 name = "cennznet-executor"
 version = "1.0.0"
 authors = ["Centrality Developers <support@centrality.ai>"]
+description = "CENNZnet node implementation in Rust."
 edition = "2018"
+license = "GPL-3.0"
+homepage = "https://cennznetdocs.com"
```
Comment (jordy25519) [inline executor/Cargo.toml]: Can remove this

File: primitives/Cargo.toml [hunk 3]
```diff
@@ -3,16 +3,19 @@ name = "cennznet-primitives"
 version = "1.0.0"
 authors = ["Centrality Developers <support@centrality.ai>"]
 edition = "2018"
+license = "GPL-3.0"
+homepage = "https://cennznetdocs.com"
```
Comment (jordy25519) [inline primitives/Cargo.toml]: Please remove

File: rpc-client/Cargo.toml [hunk 4]
```diff
@@ -3,12 +3,16 @@ name = "cennznet-rpc-client"
 version = "1.0.0"
 authors = ["Centrality Developers <support@centrality.ai>"]
 edition = "2018"
+license = "GPL-3.0"
+homepage = "https://cennznetdocs.com"
```
Comment (jordy25519) [inline rpc-client/Cargo.toml]: please remove

File: rpc/Cargo.toml [hunk 5]
```diff
@@ -3,6 +3,9 @@ name = "cennznet-rpc"
 version = "1.0.0"
 authors = ["Centrality Developers <support@centrality.ai>"]
 edition = "2018"
+license = "GPL-3.0"
+homepage = "https://cennznetdocs.com"
```
Comment (jordy25519) [inline rpc/Cargo.toml]: please remove

File: runtime/Cargo.toml [hunk 6]
```diff
@@ -4,81 +4,86 @@ version = "1.0.0"
 authors = ["Centrality Developers <support@centrality.ai>"]
 edition = "2018"
 build = "build.rs"
+license = "GPL-3.0"
+homepage = "https://cennznetdocs.com"
```
Comment (jordy25519) [inline runtime/Cargo.toml]: Please remove

File: cli/src/service.rs [hunk 7]
```diff
@@ -631,12 +615,7 @@ mod tests {
 			crate::chain_spec::kauri::config(),
 			|config| new_full(config),
 			|config| new_light(config),
-			vec![
-				"//Andrea".into(),
-				"//Brooke".into(),
-				"//Courtney".into(),
-				"//Drew".into(),
-			],
+			vec!["//Alice".into(), "//Bob".into(), "//Courtney".into(), "//Drew".into()],
```
Comment (jordy25519) [inline cli/src/service.rs]: "Andrea", "Brooke"

File: cli/src/cli.rs [hunk 8]
```diff
@@ -90,118 +74,3 @@ pub struct FactoryCmd {
...
-	type Config<A, B> = Configuration<(), A, B>;
-
-	match parse_and_prepare::<CustomSubcommands, NoCustom, _>(&version, "cennznet-node", args) {
-		ParseAndPrepare::Run(cmd) => cmd.run(
-			load_spec,
-			exit,
-			|exit, _cli_args, _custom_args, mut config: Config<_, _>| {
-				info!("{}", version.name);
-				info!("  version {}", config.full_version());
-				info!("  by Centrality Investments UNlimited, 2018-2020");
```
Comment (jordy25519) [inline cli/src/cli.rs:110]: We'll want to re-instate these log messages at some point, not sure where they are handled now

File: crml/transaction-payment/src/lib.rs [hunk 9]
```diff
@@ -325,15 +322,264 @@ where
...
+	const CALL: &<Runtime as frame_system::Trait>::Call = &Call::Balances(BalancesCall::transfer(2, 69));
+
 	// A balance transfer, which will be considered 'gas metered' for testing purposes
 	const METERED_CALL: &<Runtime as frame_system::Trait>::Call =
-		&mock::Call::Balances(pallet_balances::Call::transfer_keep_alive(GAS_METERED_ACCOUNT_ID, 69));
+		&Call::Balances(BalancesCall::transfer_keep_alive(GAS_METERED_ACCOUNT_ID, 69));
+
+	impl_outer_dispatch! {
+		pub enum Call for Runtime where origin: Origin {
+			pallet_balances::Balances,
```
Comment (jordy25519) [inline crml/transaction-payment/src/lib.rs:368]: Was there a reason to move this setup stuff out of mock.rs?

**Review by jordy25519 — APPROVED** [review]
Please remove "homepage" from Cargo.tomls
Apart from that looks good

File: crml/transaction-payment/src/lib.rs (hunk 9 above)
Comment (cowboy-bebug) [reply crml/transaction-payment/src/lib.rs:368]: The entire `mock.rs` file is removed and the contents are move here

**Review by cowboy-bebug — APPROVED** [review]
LGTM! Reinstating the startup log could be a separate task after this

File: crml/transaction-payment/Cargo.toml (hunk 1 above)
Comment (aliXsed) [reply crml/transaction-payment/Cargo.toml]: done

File: executor/Cargo.toml (hunk 2 above)
Comment (aliXsed) [reply executor/Cargo.toml]: done

File: primitives/Cargo.toml (hunk 3 above)
Comment (aliXsed) [reply primitives/Cargo.toml]: done

File: rpc-client/Cargo.toml (hunk 4 above)
Comment (aliXsed) [reply rpc-client/Cargo.toml]: done

File: rpc/Cargo.toml (hunk 5 above)
Comment (aliXsed) [reply rpc/Cargo.toml]: done

File: runtime/Cargo.toml (hunk 6 above)
Comment (aliXsed) [reply runtime/Cargo.toml]: done

File: cli/src/service.rs (hunk 7 above)
Comment (aliXsed) [reply cli/src/service.rs]: I have added these two to the list. Is that what you meant?

File: crml/transaction-payment/src/lib.rs (hunk 9 above)
Comment (aliXsed) [reply crml/transaction-payment/src/lib.rs:368]: Just to be comparable with Plug to make the possible future migration easier.

File: cli/src/cli.rs (hunk 8 above)
Comment (aliXsed) [reply cli/src/cli.rs:110]: They now seem to be coming from Plug: https://github.com/plugblockchain/plug-blockchain/blob/b7e29ddaab5a2bbf08a730424ec5dffcfcd6467f/client/cli/src/commands/runcmd.rs#L458

Comment (aliXsed) [reply cli/src/cli.rs:110]: Are you happy with what @cowboy-bebug is suggesting that we re-instate the log in a different task?

File: cli/src/service.rs (hunk 7 above)
Comment (aliXsed) [reply cli/src/service.rs]: I see Emily and Frank in some other places too. Should we add them as well?

**Comment by aliXsed** [comment]
@holygits @cowboy-bebug This branch at this stage builds and tests okay. I'm now ready to do the migration to Plug rc3.

File: crml/staking/Cargo.toml [hunk 10]
```diff
@@ -3,28 +3,31 @@ name = "crml-staking"
...
-pallet-balances = { git = "https://github.com/plugblockchain/plug-blockchain", branch = "1.0.0-rc2" }
-pallet-generic-asset = { git = "https://github.com/plugblockchain/plug-blockchain", branch = "1.0.0-rc2" }
-pallet-timestamp = { git = "https://github.com/plugblockchain/plug-blockchain", branch = "1.0.0-rc2" }
-crml-staking-reward-curve = { version = "1.0.0",  path = "../staking/reward-curve/" }
-substrate-test-utils = { git = "https://github.com/plugblockchain/plug-blockchain", branch = "1.0.0-rc2" }
+sp-core = { git = "https://github.com/plugblockchain/plug-blockchain", branch = "develop" }
+pallet-balances = { git = "https://github.com/plugblockchain/plug-blockchain", branch = "develop" }
+pallet-generic-asset = { git = "https://github.com/plugblockchain/plug-blockchain", branch = "develop" }
+pallet-timestamp = { git = "https://github.com/plugblockchain/plug-blockchain", branch = "develop" }
+crml-staking-reward-curve = {  path = "../staking/reward-curve" }
```
Comment (jordy25519) [inline crml/staking/Cargo.toml]: whitespace `  path`

**Comment by cowboy-bebug** [comment]
> @holygits @cowboy-bebug This branch at this stage builds and tests okay. I'm now ready to do the migration to Plug rc3.

Nice work! Hopefully the migration is less painful now 😄

File: cli/src/service.rs [hunk 11]
```diff
@@ -632,10 +616,12 @@ mod tests {
 			|config| new_full(config),
 			|config| new_light(config),
 			vec![
-				"//Andrea".into(),
-				"//Brooke".into(),
+				"//Alice".into(),
```
Comment (jordy25519) [inline cli/src/service.rs]: Alice and Bob should be replaced with Andrea and Brooke here

**Comment by aliXsed** [comment]
> LGTM! Reinstating the startup log could be a separate task after this

@holygits @cowboy-bebug  I just noticed we are already providing the log info here:  https://github.com/cennznet/cennznet/blob/9a6df4422a3f527a27b76d6273b117e9151fb095/src/main.rs#L25
We probably just need to test it.

File: crml/staking/Cargo.toml (hunk 10 above)
Comment (aliXsed) [reply crml/staking/Cargo.toml]: done

File: runtime/tests/tests.rs [hunk 12]
```diff
@@ -679,6 +679,7 @@ fn contract_call_fails_with_insufficient_gas_with_fee_exchange() {
 		});
 }
 
+#[ignore]
```
Comment (jordy25519) [inline runtime/tests/tests.rs]: We can un-ignore these now and can update the fee amounts in the tests to match.
Would be good to know the reason for the change but not a blocker for this PR

File: runtime/build.rs [hunk 13]
```diff
@@ -14,14 +14,13 @@
...
-	build_current_project_with_rustflags(
-		"wasm_binary.rs",
-		WasmBuilderSource::Crates("1.0.8"),
-		// This instructs LLD to export __heap_base as a global variable, which is used by the
-		// external memory allocator.
-		"-Clink-arg=--export=__heap_base",
-	);
+	WasmBuilder::new()
+		.with_current_project()
+		.with_wasm_builder_from_git("https://github.com/plugblockchain/plug-blockchain", "develop")
```
Comment (jordy25519) [inline runtime/build.rs]: "develop" -> "1.0.0-rc3"

File: runtime/tests/tests.rs (hunk 12 above)
Comment (aliXsed) [reply runtime/tests/tests.rs]: done

File: runtime/build.rs (hunk 13 above)
Comment (aliXsed) [reply runtime/build.rs]: done

**Pull request merged**